src/components/OfflineSearch.ts

```diff
@@ -0,0 +1,173 @@
+        atEnd: () => boolean;
+        _pageSize: number;
+    };
+    _loadMore: () => void;
+    _onLoad: () => void;
+    _renderData: () => void;
+    update: () => void;
+}
+
+interface OfflineSearchProps extends Props<OfflineSearch> {
+    defaultQuery: string;
+    searchEntity: string;
+    searchAttribute: string;
+    showSearchBar: string;
+    targetGridName: string;
+    searchMethod: searchMethodOptions;
+}
+
+interface OfflineSearchState {
+    alertMessage?: string;
+    buttonVisibility?: string;
+}
+
+class OfflineSearch extends Component<OfflineSearchProps, OfflineSearchState> {
+    // internal variables
+    private searchButton: HTMLButtonElement;
+    private searchInput: HTMLInputElement;
+    private targetWidget: ListView;
+
+    constructor(props: OfflineSearchProps) {
+        super(props);
+
+        this.state = {
+            alertMessage: "",
+            buttonVisibility: "hidden"
+        };
+        this.onClear = this.onClear.bind(this);
+        this.onSearchKeyDown = this.onSearchKeyDown.bind(this);
+    }
+
+    render() {
+        return createElement("div", {
+            className: `widget-offline-search ${this.props.showSearchBar}`
+        },
+            createElement(Alert, { message: this.state.alertMessage }),
+            createElement("div", { className: "search-container" },
+                createElement("span", { className: "glyphicon glyphicon-search" }),
+                createElement("input", {
+                    className: "form-control", placeholder: "Search",
+                    ref: input => this.searchInput = input as HTMLInputElement
+                }),
+                createElement("button", {
+                    className: `btn-transparent ${this.state.buttonVisibility}`,
+                    ref: button => this.searchButton = button as HTMLButtonElement
+                },
+                    createElement("span", { className: "glyphicon glyphicon-remove" })
+                )
+            )
+        );
+    }
+
+    componentDidMount() {
+        this.findTargetNode();
+        this.isValidWidget();
+        this.setUpEvents();
+        this.setDefaultSearch();
+    }
+
+    componentWillUnmount() {
+        this.searchButton.removeEventListener("click", this.onClear);
+        this.searchInput.removeEventListener("keyup", this.onSearchKeyDown);
+    }
+
+    private findTargetNode() {
+        let queryNode = findDOMNode(this).parentNode as HTMLElement;
+        let targetNode: HTMLElement | null = null;
+        const targetName = this.props.targetGridName;
+
+        while (!targetNode) {
+            targetNode = queryNode.querySelector(`.mx-name-${targetName}`) as HTMLElement;
+            if (window.document.isEqualNode(queryNode)) break;
+            queryNode = queryNode.parentNode as HTMLElement;
+        }
+
+        if (!targetNode) {
+            this.setState({ alertMessage: `search offline widget: unable to find grid with the name "${targetName}"` });
```

Comment:
I really dont see the point of moving that part to the `componentDidMount` yet already the function that contains the code is being called.